transcom/mymove
# INT B-19237 customer cac registration

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-19237)
[Feature Flag](https://github.com/transcom/milmove-feature-flags/pull/111)

## Summary

We are wanting the customer to authenticate with their smart card at least ONCE. We do not want them to set up a move  or their profile until they have logged in and authenticated with CAC.

When a user is authenticating, we have access to the type of authenticators they used to login. In order to track if they've EVER logged in with CAC, we needed to add a column to the `service_members` table in the database called `cac_validated`, which is set to `false` by default.

When a new user is created, we will leave that value as `false` and then have logic checks when authorizing a known user to see if they're:
1. Signing in with CAC
2. Are accessing the customer app
3. Is the value of `cac_validated` set to `false`

If those are all true, then we want to update that value so they can now start a move.

This PR does the following:
- adds migration file to add `cac_validated` column to the `service_members` table
- updates `OktaSessionInfo` struct to store if they're signing in with a smart card
- added a function, `didUserSignInWithSmartCard` to check the `amr` value in the `IDToken`
- logic check to handle turning `cac_validated` to `true` when signing in as a customer
- updated `devlocal` process to turn `cac_validated` to `true` to satisfy check when testing/developing
- updated `ServiceMember` struct to include `CacValidated`
- lotta playwright test updates
- `SmartCardRedirect` component
- set up feature flag to turn this logic on/off
- set up routing rules to handle showing the `SmartCardRedirect` screen when they have not validated with cac

### How to test

1. Before firing up the server, ensure `FEATURE_FLAG_CAC_VALIDATED_LOGIN` is set to `true` in your `envrc`
2. Fire it up as well as the client
3. Sign up as a customer and register for an Okta account. You can put in a bogus EDIPI for now
4. You'll be redirected back to Okta and you should see the below screenshot. 
5. Sign out.
6. Now in order to satisfy this check, you'll need to change the value of `cac_validated` to `true` in the db
- One way to do this is manually
- The other way would be to get with me to change your actual Okta profile EDIPI value from your current profile to the customer profile you created. Don't mind. Just hit me up.
7. Once that value is updated, sign in again with CAC or normally (if you updated it manually)
8. Should now be able to create a move

## Screenshots
Smart card info screen
![Screenshot 2024-03-28 at 8 00 24 AM](https://github.com/transcom/mymove/assets/136510600/4784f43f-82fe-476e-9327-848a66c1486d)
